bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -1090,28 +1106,46 @@ private Header getHeaderForLogId(long entryLogId) throws IOException {
             return new Header(headerVersion, ledgersMapOffset, ledgersCount);
         } finally {
             headers.release();
+            if (bc != null) {
+                bc.release();
+            }
         }
     }
 
-    private BufferedReadChannel getChannelForLogId(long entryLogId) throws IOException {
-        BufferedReadChannel fc = getFromChannels(entryLogId);
-        if (fc != null) {
-            return fc;
-        }
-        File file = findFile(entryLogId);
-        // get channel is used to open an existing entry log file
-        // it would be better to open using read mode
-        FileChannel newFc = new RandomAccessFile(file, "r").getChannel();
-        FileChannel oldFc = logid2FileChannel.putIfAbsent(entryLogId, newFc);
-        if (null != oldFc) {
-            newFc.close();
-            newFc = oldFc;
-        }
-        // We set the position of the write buffer of this buffered channel to Long.MAX_VALUE
-        // so that there are no overlaps with the write buffer while reading
-        fc = new BufferedReadChannel(newFc, conf.getReadBufferBytes());
-        putInReadChannels(entryLogId, fc);
-        return fc;
+    /**
+     * Add one refCnt for BufferedReadChannel and return it, caller need to subtract one refCnt.
+     * @param entryLogId
+     * @return
+     * @throws IOException
+     */
+    private EntryLogBufferedReadChannel getChannelForLogId(long entryLogId) throws IOException {
+        try {
+            EntryLogBufferedReadChannel brc;
+            Callable<EntryLogBufferedReadChannel> loader = () -> {
+                CachedFileChannel cachedFileChannel = fileChannelBackingCache.loadFileChannel(entryLogId);
+                // We set the position of the write buffer of this buffered channel to Long.MAX_VALUE
+                // so that there are no overlaps with the write buffer while reading
+                return new EntryLogBufferedReadChannel(cachedFileChannel, conf.getReadBufferBytes());
+            };
+            do {
+                // Put the logId, bc pair in the cache responsible for the current thread.
+                brc = getThreadLocalCacheForReadChannel().get(entryLogId, loader);
+                if (!brc.cachedFileChannel.tryRetain()) {
```

Comment:
better to do this in the loader:

```
() -> {
    while (true) {
        CachedFileChannel cachedFileChannel = fileChannelBackingCache.loadFileChannel(entryLogId);
        if (cachedFileChannel.tryRetain()) {
            return new EntryLogBufferedReadChannel(cachedFileChannel, conf.getReadBufferBytes());
        }
    }
}
```

You'll avoid having to add and remove from cache.

Reply:
Ah yes, sorry, you're right. This is the same as the defensive code in IndexPersistentMgr to avoid a bug that caused a tight loop.